Bugfix #3054 main_v12.0 parusr

JohnHalleyGotway opened the pull request:
## Expected Differences ##

This PR fixes two bugs:
1. PB2NC abort from BUFRLIB with `PARUSR` error message when pb2nc is compiled with `-O2` on Intel compilers and `obs_bufr_var` is set to an empty list in the configuration file. Please see this [issue comment](https://github.com/dtcenter/MET/issues/3054#issuecomment-2623020214) for an explanation of the problem and the fix. 
2. Separately, I propose updating `pair_base.cc` to fix a logging bug that arose during METplus/discussions#2875. The `calc_obs_summary()` function was updating the `summary_val` in a local copy rather than the value in the map itself which led to erroneous summary values being reported in the log output.

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Debugged and tested extensively at the `met_test` user in `seneca:/d1/projects/MET/intel/parusr_bufrlib_error`.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
As the `met_test` user, run the following command to see that the (1) METv12.0.0 release aborts while the (2) `bugfix_3054_main_v12.0_parusr` and (3) `bugfix_3054_develop_parusr` branches run to completion (when compiled with Intel and `-O2`).
```
runas met_test
cd /d1/projects/MET/intel/parusr_bufrlib_error
source /d1/projects/MET/intel/MET-bugfix_3054_develop_parusr/internal/scripts/environment/development.seneca_intel_oneapi
# Confirm I compiled with -O2
egrep "CFLAGS=|FFLAGS=|CXXFLAGS=" /d1/projects/MET/intel/MET-*/config.log
./run_pb2nc_test.sh
```
- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**
None needed.

- [x] Do these changes include sufficient testing updates? **[No]**
None needed.

- [x] Will this PR result in changes to the MET test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Will this PR result in changes to existing METplus Use Cases? **[No]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.

- [x] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:

- [x] Please complete this pull request review by **[Fri 1/31/25]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

Commit:
Per #3054, fix PARUSR BUFRLIB error by solving the upstream reference to temporary memory returned by c_str(). Store a copy of the temporary variable name in a string rather than a pointer to temporary memory. Note that I checked all other calls to c_str() in pb2nc.cc and found these 2 instances to be only problematic ones. All others are used as arguments to functions for which a copy is made.

Commit:
Unrelated to #3054, but discovered while investigating the dtcenter/METplus#2875 discussion, the PairBase::calc_obs_summary() function loops over map entries and attempts to update the mapped 'summary_val' value. However, the current version only updates it in a copy and not what's actually in the map. This changes how we loop over the map to actually udpate its contents. Note that the only impact is fixing a log file to accurately report the 'summary_val'. So this is really a logging bug.

georgemccabe commented:
@JohnHalleyGotway, I see that this run flagged differences. The checklist notes that there should not be any differences. Could you take a look?

georgemccabe reviewed (changes requested):
I was able to get a successful run of the failing pb2nc unit test using this fix. I had already set up an environment on seneca to replicate the failure and confirmed that applying this fix resolved the issue.

I do see some differences that were flagged that appear to be unexpected. Once those are resolved/explained, I approve. Thanks for fixing this!

JohnHalleyGotway commented:
Thanks @georgemccabe for the review. There was a big problem with the bugfix I made to `pair_base.cc`. I have corrected and ran the following commands on seneca to demonstrate:

- This script runs the existing main_v12.0 version of Point-Stat followed by the bugfix version for the 2 commands in which diffs were flagged:
```
cd /d1/projects/MET/intel/MET-bugfix_3054_main_v12.0_parusr/internal/test_unit/
./run_ps.sh
```

- Diff the log output from existing/bugfix runs to show that bogus summary values reported in the log output are now corrected:
```
220c220
< DEBUG 4: Computed NEAREST of 1 observations = 6.95286e-310 for [lat:lon:level:elevation] = [55.280:-114.780:-9999.00:-9999.00]
---
> DEBUG 4: Computed NEAREST of 1 observations = -1.7 for [lat:lon:level:elevation] = [55.280:-114.780:-9999.00:-9999.00]
222c222
< DEBUG 4: Computed NEAREST of 2 observations = 6.95286e-310 for [lat:lon:level:elevation] = [49.030:-122.360:-9999.00:-9999.00]
---
> DEBUG 4: Computed NEAREST of 2 observations = -1.8 for [lat:lon:level:elevation] = [49.030:-122.360:-9999.00:-9999.00]
```

- Compare the existing/bugfix output files to show there are no diffs:
```
for file in `ls $MET_TEST_OUTPUT/point_stat/point_stat_SID_INC_EXC*`; do
  base=`basename $file`; echo $base `diff $file /d1/projects/MET/MET_regression/main_v12.0/NB20250131/MET-main_v12.0/test_output/point_stat/${base} | wc -l`;
done
point_stat_SID_INC_EXC_120000L_20120409_120000V_mpr.txt 0
point_stat_SID_INC_EXC_120000L_20120409_120000V.stat 0
point_stat_SID_INC_EXC_CENSOR_120000L_20120409_120000V_mpr.txt 0
point_stat_SID_INC_EXC_CENSOR_120000L_20120409_120000V.stat 0
```

Commit:
Per #3054, revert emplace_back() to its original push_back() to make the bugfix diffs as limited as possible.

Commit:
Per #3054, correct bugfix in PairBase::calc_obs_summary() in pair_base.cc

georgemccabe reviewed (approved):
Looks good
